c/tskit/genotypes.h

```diff
@@ -34,46 +34,150 @@ extern "C" {
-    tsk_id_t *alt_samples; /* alternative subset of samples to use */
+    tsk_id_t *alt_samples;
     tsk_id_t *alt_sample_index_map;
 
 } tsk_variant_t;
 
+/* All vargen related structs and methods were deprecated in C API v1.0 */
 typedef struct {
     const tsk_treeseq_t *tree_sequence;
     tsk_id_t site_index;
     tsk_variant_t variant;
 } tsk_vargen_t;
 
+/**
+@defgroup VARIANT_API_GROUP Variant API for obtaining genotypes.
+@{
+*/
+
+/**
+@brief Initialises the variant by allocating the internal memory and
+initialising the internal tree
+
+@rst
+This must be called before any operations are performed on the variant.
+See the :ref:`sec_c_api_overview_structure` for details on how objects
+are initialised and freed.
+@endrst
+
+@param self A pointer to an uninitialised tsk_variant_t object.
+@param tree_sequence A pointer to the tree sequence from which this variant
+will decode genotypes. No copy is taken, so this tree sequence must persist
+for the lifetime of the variant.
+@param samples An array of node ids of the samples that are to have their genotypes
+decoded. A copy of this array will be taken by the variant.
+@param num_samples The number of ids in the samples array.
+@param alleles Optional. Either ``NULL`` or an array of string alleles with a terminal
+``NULL`` sentinel value.
+If specified, the genotypes will be decoded to match the index in this allele array.
+If ``NULL`` then alleles will be automatically determined from the mutations encountered.
+@param options Variant options. Either ``0`` or ``TSK_ISOLATED_NOT_MISSING`` which if
+specifed indicates that isolated sample nodes should not be decoded as the "missing"
+state but as the ancestral state (or the state of any mutation above them).
+@return Return 0 on success or a negative value on failure.
+*/
 int tsk_variant_init(tsk_variant_t *self, const tsk_treeseq_t *tree_sequence,
     const tsk_id_t *samples, tsk_size_t num_samples, const char **alleles,
     tsk_flags_t options);
+
+/**
+@brief Copys the state of this variant to another variant
+
+@rst
+Copies the site, genotypes and alleles from this variant to another. Note that
+the other variant should be uninitialised as this method does not free any
+memory that the other variant owns. The tree is not copied to the new variant
+this restricts it from being further decoded at any site, and is frozen to the site
+of the copied variant.
+@endrst
+
+@param self A pointer to an initialised and decoded tsk_variant_t object.
+@param other A pointer to an uninitialised tsk_variant_t object.
+@return Return 0 on success or a negative value on failure.
+*/
 int tsk_variant_restricted_copy(tsk_variant_t *self, tsk_variant_t *other);
-int tsk_variant_decode(
-    tsk_variant_t *self, tsk_id_t site_id, tsk_flags_t TSK_UNUSED(options));
+
+/**
+@brief Decode the genotypes at the given site, storing them in this variant.
+
+@rst
+Uses the internal tree of this variant to seek to the tree of the specified site
```

Comment:
Let's keep the implementation details hidden. We can just say that this is most efficient if done sequentially.

Reply:
https://github.com/tskit-dev/tskit/pull/2186/commits/d210fce3798756d45576b456a227b769efff6a6d